dace/transformation/passes/constant_propagation.py

```diff
@@ -236,12 +369,67 @@ def _add_nested_datanames(name: str, desc: data.Structure):
                         if reassignments and (used_in_assignments - reassignments):
                             assignments[aname] = _UnknownValue
 
-                if state not in result:  # Condition may evaluate to False when state is the start-state
-                    result[state] = {}
-                redo |= self._propagate(result[state], assignments)
-
-        return result
-
+                if isinstance(block, LoopRegion):
+                    # Any constants before a loop that may be overwritten inside the loop cannot be assumed as constants
+                    # for the loop itself.
+                    assigned_in_loop = self._assignments_in_loop(block)
+                    for k in assignments.keys():
+                        if k in assigned_in_loop:
+                            assignments[k] = _UnknownValue
+
+                if block not in in_const_dict:
+                    in_const_dict[block] = {}
+                if assignments:
+                    redo |= self._propagate(in_const_dict[block], assignments)
+
+                if isinstance(block, ControlFlowRegion):
```

Comment:
shouldn't the conditions be inverted? i.e., ConditionalBlock first and ControlFlowRegion second? Isn't one a subclass of the other?

Reply:
No, that's part of the problem that `AbstractControlFlowRegion`s solve.